core/src/main/java/smile/manifold/UMAP.java

```diff
@@ -0,0 +1,559 @@
+		int[][] neighbors = new int[data.length][n];
+		Graph knns = nearest_neighbors(data, n, neighbors);
+		double[] sigmas = new double[data.length];
+		double[] rhos = new double[data.length];
+		SparseMatrix coo = fuzzy_simplicial_set(knns, n, sigmas, rhos);
+
+		// Spectral embedding for initialization
+		coordinates = spectral_layout(coo, d);
+		
+		// Optimizing the embedding
+		optimize_layout(coordinates, null, 1);
+		
+	}
+
+	/**
+	 * general curve to be fit:
+	 * 
+	 * <pre>
+	 *    1.0 / (1.0 + a * x ** (2 * b))
+	 * </pre>
+	 */
+	private static DifferentiableMultivariateFunction func = new DifferentiableMultivariateFunction() {
+
+		@Override
+		public double f(double[] x) {
+			return 1 / (1 + x[0] * Math.pow(x[2], x[1]));
+		}
+
+		@Override
+		public double g(double[] x, double[] g) {
+			double pow = Math.pow(x[2], x[1]);
+			double de = 1 + x[0] * pow;
+			g[0] = -pow / (de * de);
+			g[1] = -(x[0] * x[1] * Math.log(x[2]) * pow) / (de * de);
+			return 1 / de;
+		}
+	};
+
+	/**
+	 * Fit a, b params for the differentiable curve used in lower dimensional
+	 * fuzzy simplicial complex construction. We want the smooth curve (from a
+	 * pre-defined family with simple gradient) that best matches an offset
+	 * exponential decay.
+	 * 
+	 * @param spread
+	 *            The effective scale of embedded points. In combination with
+	 *            minDist, this determines how clustered/clumped the embedded
+	 *            points are.
+	 * @param minDist
+	 *            The desired separation between close points in the embedding
+	 *            space. The value should be set no-greater than and relative to
+	 *            the spread value, which determines the scale at which embedded
+	 *            points will be spread out
+	 * @return a, b params for the differentiable curve
+	 */
+	private static double[] find_ab_params(double spread, double minDist) {
+		int xSize = 300;
+		double[] x = new double[xSize];
+		double[] y = new double[xSize];
+		double xEnd = 3 * spread;
+		double xInterval = xEnd / (xSize - 1);
+		for (int i = 0; i < x.length; i++) {
+			x[i] = i * xInterval;
+			y[i] = x[i] < minDist ? 1 : Math.exp(-(x[i] - minDist) / spread);
+		}
+		double[] p = {1, 1};
+		LevenbergMarquardt curveFit = LevenbergMarquardt.fit(func, x, y, p);
+		return new double[]{curveFit.p[0], curveFit.p[1]};
+	}
+	
+	/**
+	 * Compute the n nearest points for each data point using approximate
+	 * searching for speed consideration
+	 * 
+	 * @param data
+	 *            input data for which to search nearest neighbors.
+	 * @param n
+	 *            The number of neighbors to consider when approximating the
+	 *            neighbors.
+	 * @param neighbors
+	 *            record to hold nearest neighbors for each data point
+	 * @return the n nearest points for each data point in a
+	 *         {@link NearestNeighborGraph}
+	 */
+	private static Graph nearest_neighbors(double[][] data, int n,
```

Comment:
I make a small change to `NearestNeighbor.of()` by removing 'Optional`, which should only be used for return values.